pkg/apis/pipeline/v1beta1/pipeline_types.go

```diff
@@ -185,6 +202,88 @@ func (pt PipelineTask) ValidateName() *apis.FieldError {
 	return nil
 }
 
+// Validate custom task specifications - checking kind and fail if not yet supported features specified
+func (pt PipelineTask) ValidateCustomTask() (errs *apis.FieldError) {
+	errs = errs.Also(pt.validate())
+	if pt.TaskRef.Kind == "" {
+		errs = errs.Also(apis.ErrInvalidValue("custom task ref must specify kind", "taskRef.kind"))
+	}
+	// Conditions are deprecated so the effort to support them with custom tasks is not justified.
+	// When expressions should be used instead.
+	if len(pt.Conditions) > 0 {
+		errs = errs.Also(apis.ErrInvalidValue("custom tasks do not support conditions - use when expressions instead", "conditions"))
+	}
+	// TODO(#3133): Support these features if possible.
+	if pt.Retries > 0 {
+		errs = errs.Also(apis.ErrInvalidValue("custom tasks do not support retries", "retries"))
+	}
+	if pt.Resources != nil {
+		errs = errs.Also(apis.ErrInvalidValue("custom tasks do not support PipelineResources", "resources"))
+	}
+	if pt.Timeout != nil {
+		errs = errs.Also(apis.ErrInvalidValue("custom tasks do not support timeout", "timeout"))
+	}
+	return errs
+}
+
+// Validate bundle specifications - checking name and bundle
+func (pt PipelineTask) ValidateBundle() (errs *apis.FieldError) {
+	errs = errs.Also(pt.validate())
+	// bundle requires a TaskRef to be specified
+	if (pt.TaskRef != nil && pt.TaskRef.Bundle != "") && pt.TaskRef.Name == "" {
+		errs = errs.Also(apis.ErrMissingField("taskRef.name"))
+	}
+	// If a bundle url is specified, ensure it is parsable
+	if pt.TaskRef != nil && pt.TaskRef.Bundle != "" {
+		if _, err := name.ParseReference(pt.TaskRef.Bundle); err != nil {
+			errs = errs.Also(apis.ErrInvalidValue(fmt.Sprintf("invalid bundle reference (%s)", err.Error()), "taskRef.bundle"))
+		}
+	}
+	return errs
+}
+
+// Validate regular pipeline task (dag/final) for taskRef and taskSpec
+func (pt PipelineTask) ValidateRegularTask(ctx context.Context) (errs *apis.FieldError) {
```

Comment:
We don't use it anywhere as of today, so we could keep this one unexported.

Reply:
yup, changed it to unexported.